src/main/java/com/vlkan/pubsub/util/BoundedScheduledThreadPoolExecutor.java

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright 2019 Volkan Yazıcı
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permits and
+ * limitations under the License.
+ */
+
+package com.vlkan.pubsub.util;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.*;
+
+/**
+ * A {@link ScheduledThreadPoolExecutor} wrapper enforcing a bound on the task
+ * queue size. Excessive task queue growth yields {@link
+ * RejectedExecutionException} errors. {@link RejectedExecutionHandler}s are
+ * not supported since they expect a {@link ThreadPoolExecutor} in their
+ * arguments.
+ *
+ * <p>Java Standard library unfortunately doesn't provide any {@link
+ * ScheduledExecutorService} implementations that one can provide a bound on
+ * the task queue. This shortcoming is prone to hide backpressure problems. See
+ * <a href="http://cs.oswego.edu/pipermail/concurrency-interest/2019-April/016861.html">the
+ * relevant concurrency-interest discussion</a> for {@link java.util.concurrent}
+ * lead Doug Lea's tip for enforcing a bound via {@link
+ * ScheduledThreadPoolExecutor#getQueue()}.
+ */
+@SuppressWarnings("NullableProblems")
+public class BoundedScheduledThreadPoolExecutor implements ScheduledExecutorService {
```

Comment:
Would it work if Reactor had a special scheduler, something like `elastic` but with a maximum size of the pool and a queue in front?

Reply:
That would be awesome -- not an `elastic`, but a `parallel` one. I cannot understand how come people can run production systems with schedulers where the thread pool and task queue are unbounded. Whenever I have introduced this simple class to a colleague, they returned back to me with their cursings (since it exposed a shit load of backpressure problems in the code) and blessings (since it exposed a shit load of backpressure problems in the code). :smile:

In an ideal reactive system, you shouldn't need a thread pool bigger than the available number of cores. If you have I/O- and processing-sensitive tasks, you can shepherd them under different schedulers. And an unbounded task queue is... just evil.

Reply:
FYI https://github.com/reactor/reactor-core/pull/1881

Reply:
Doesn't this just cap the pool size whereas I was expecting a cap at the task queue size?

Reply:
I guess it is a good timing to chime in and request "queue size" parameter :)